documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -834,6 +831,103 @@ public interface IJoystickInputHandler : IInputHandler
 
 `HandleHatMove` must be called when any value of `JoystickState.Hats` changes.
 
+# Input Maps
+
+## Presentation
+Input maps are designed to be easyer to the user to handle actions with more than one input.
+At first view it can looks useless and a bad practice, but this is complemtelly useful for cross-plataform programs, being games an example.
+
+This is an pseudocode example of what the Input Map system solves:
+
+Without inputMaps:
+```cs
+inputContext.Update();
+
+private Dictionary<string, bool> pressedKeys = new();
+
+inputContext.Gamepads.ThumbstickMove += @event => {Move(@event.Value)};
+inputContext.Joysticks.AxisMove += @event => {Move(@event.Value)};
+
+private Move(Vector2D<float> value)
+{
+    // Code for movement here
+}
+```
+
+With inputMaps:
+```cs
+inputContext.Update();
+
+// This single line get rid of the two in the example above. TWO ISN'T THE LIMIT!
+inputContext.Map.GetAction("Move").Axis += @event => {Move(@event.Value)};
+
+private Move(Vector2D<float> value)
+{
+    // Code for movement here
+}
+```
+
+## Implementation
+Each InputContext have one reference to an InputMap.
+This is how the InputMap class work:
+```cs
+public partial class InputMap
+{
+    public InputMapAction GetAction(string name);
+    public void AddAction(InputMapAction action, string name);
+    public void RemoveAction(string name);
+    public static InputMap Load(string path);
+    public static void Save(InputMap source, string path);
+    public static string Compile(InputMap source);
+}
+```
+
+`GetAction` method is the main entry point of the system. It allows the use to get an mapped action by it configurated name.
+
+```cs
+public partial class InputMapAction
+{
+    public bool IsPressed { get; }
+    public event Action<AxisInputEvent> Axis;
+}
+```
+
+## Configuration and saving
+The input map system's save system is one of the most important to it working. It helps to make the configuration easyer as faster.
+
+Instead of needing to hardcode all the actions, the user should create a file with a special syntax. This is an example of the sintax of the map system's saved data:
+```
+"move_up":
+    - KeyName.Up
+    - JoystickButton.DPadUp
+
+"move_down":
+- KeyName.Down
+- JoystickButton.DPadDown
+
+"move":
+    - Vector2D [
+        -KeyName.Down
+        -KeyName.Up
+        -KeyName.Left
+        -KeyName.Right
+    ]
+    - Joystick.Axis
+```
+
+The main syntax may somewhat resemble YAML, but it's much more symple and easy to parse.
```

Comment:
Why are we making our own file format? This is scope creep and beyond our engineering resources to execute. Using a well-supported format such as JSON is perfectly fine, a custom format is over engineering.

Reply:
i personally would lean towards ini due to its simplicity for non-developers to read/write.
Although the question is, are end users (eg. a player of a game built on Silk.NET) ever intended to modify this file? Or is this mapping file for developers only? The API being hardcoded to read/write to a path implies to me that its meant to be handled by end-users aswell, and JSON can be quite daunting to newcomers

Reply:
I don't want to use existing file formats like json or YAML (that is a base for syntax) because an entire parser for these ones will be unnecessarily expensive for something so simple. Also i want a file format easy for anyone to edit, being developers or users. Also i want something easy and fast to read and parse

This file format can use just a single loop to iterate per line and the state machine-based structure makes you and the machine easily understand the scopes

So i think it can be more optimized and cheaper if we do our own file format that does JUST what we want with an easy to parse syntax instead of use a complex data structure that needs a lot of operations to parse

Reply:
I can guarantee that System.Text.Json will be infinitely faster than any parser that the Silk.NET team has the ability to maintain itself. The same is likely true for extant YAML parsers.

This proposal creates unnecessary maintenance effort for the Silk.NET team that the team won't want to be dealing with, and will likely refuse to implement or merge on those grounds.

Reply:
understand. I will search about the YAML and JSON parsers and if you're right in this case, i will change to one of these two